### PR TITLE
Add meson.superproject()

### LINKED ISSUE #2844: Accessing top-level project variables from subproject
Is it possible to re-use the dependency declared in the top-level project from the subproject?

#### Motivation

My main project depends on BLAS library and its subproject (say, 'MUMPS') also depends on BLAS.
There are multple BLAS implementations around, but I want to use the same BLAS dependency in both.

It is already working with if BLAS is provided by a subproject (say, 'coinblas'), then I can use `subproject('coinblas').get_variable('BLAS_dep')` from 'MUMPS' subproject.
What I want is `superproject().get_variable('BLAS_dep')`.

### PR DESCRIPTION
**Pull request by alyst**
Adds `meson.superproject()` method that returns top-level project object.
The superproject object supports 2 methods:
  * `get_variable()`: access to the top-level project variables (defined by the time of subproject `run()` invocation), same as `subproject.get_variable()`
  * `get_option()`: access to the options of the top-level project

Resolves #2844

### CODE REVIEW

**Comment by nirbheek** [comment]
This does not make any sense, so it will not be added. Meson subprojects are meant to be self-contained projects. If a project requires a 'super-project', it can no longer be built by itself.

If you have a set of build files that build a dependency that will *never* be built by itself, build it as a subdir. You can then use git submodules or something like that to share it between projects.

**Comment by alyst** [comment]
In #2844 I describe what I want to achieve.
There's already `is_subproject()` method, so apparently a project can have different behaviour depending whether it's a standalone or a subproject.
So if the project is not a subproject or its super-project doesn't provide the dependency, it will just define an alternative way to get this dependency.

**Comment by nirbheek** [comment]
> In #2844 I describe what I want to achieve.

And this PR describes one solution for what you want to achieve, and I just explained why it doesn't work. I describe the preferred solution on that issue. Note that you are not a unique situation, it's been solved hundreds of times in the way I described on that issue.

A subproject *cannot* know what the parent-project's structure is, what the variable names are, and so on, so it must never depend on its structure. I mean, sure, if you handle both the subproject and the parent project and no one else ever uses the subproject, it'll work. But you do see how that defeats the whole point of the subproject feature, right?

To put it another way, circular dependencies on the internal structure of projects leads to a situation worse than DLL-hell. In fact, the very idea that a subproject will ever have one single 'super-project' is a lie. Subprojects very often use variables from each other, and what was once a parent-project can be a subproject in another configuration. For example:

You're building gtk+, which depends on glib. gtk+ is the 'superproject', glib is a subproject.

You're building Hexchat, which depends on gtk+ and glib. Hexchat is the 'superproject' and gtk+ and glib are subprojects.

**Comment by alyst** [comment]
I'm not insisting on including this PR at all :)
It was simple to implement (simpler than `dependency('blas')`), I just decided to share it with the community.

It's true that with `superproject()` people would be more tempted to define projects that rely on some hierarchical structure.
But it's easy to reject such projects from the public WrapDB. Whether it's a bad idea even for custom dedicated WrapDB repositories, I just don't know.

**Comment by nirbheek** [comment]
Any feature that can be misused, will be misused, so we have to be careful in what to add. Wrapdb is only for projects that have upstreams which probably won't accept meson build files.

Closing. Do open a new PR for the blas dependency implementation if you end up working on it, cheers.

**Comment by jpakkane** [comment]
One thing that we can do which I have been thinking is make it possible for a project option to "yield" to a one in the superproject. So something like:

    option('foo', ... , yield_to_superproject : true)

Then if this project is being built as a subproject, _and_ if the superproject has an option also called `foo`, then doing `get_option('foo')` in the subproject would transparently return the value of the superproject. In all other cases it would return its own value.

This is for cases where you have a (usually legacy) project with many disconnected git repos but still want to have a project option that has the same value in all of them and which you can switch with a single call to `meson configure`.

**Comment by tp-m** [comment]
@jpakkane: yes please, I'd really like that as well. Not just for legacy projects with disconnected git repos, but it's also useful for e.g. GStreamer's `gst-build`, so you can enable/disable gobject-introspection or docs building or such with one single switch for all subprojects (plugin modules).

**Comment by nirbheek** [comment]
> One thing that we can do which I have been thinking is make it possible for a project option to "yield" to a one in the superproject

I've wanted this for a while too, and to bikeshed the name we can call it `inherit: true`.

I also think this should be the accompanied by a convention on what to call your options for common actions such as enable/disable generation of documentation, introspection, man pages, tests, and so on, 

I also think the default should be `inherit: true` to avoid boilerplate in every project to make it work properly in subprojects.
